app/models/manageiq/providers/vmware/infra_manager/inventory/collector.rb

```diff
@@ -19,108 +19,78 @@ def run
 
     saver.start_thread
 
+    version = initial_refresh
+    _log.info("Refreshing initial inventory...Complete")
+
     until exit_requested
-      monitor_updates
+      version = monitor_updates(version)
     end
 
     saver.stop_thread
 
     _log.info("Monitor updates thread exited")
+  ensure
+    destroy_property_filter
+    disconnect
   end
 
   def stop
     _log.info("Monitor updates thread exiting...")
     self.exit_requested = true
   end
 
-  def monitor_updates
-    vim = connect
-    wait_for_updates(vim)
-  ensure
-    disconnect(vim)
-  end
-
-  private
-
-  def connect
-    host = ems.hostname
-    username, password = ems.auth_user_pwd
-
-    _log.info("Connecting to #{username}@#{host}...")
-
-    vim_opts = {
-      :ns       => 'urn:vim25',
-      :host     => host,
-      :ssl      => true,
-      :insecure => true,
-      :path     => '/sdk',
-      :port     => 443,
-      :rev      => '6.5',
-    }
-
-    require 'rbvmomi/vim'
-    conn = RbVmomi::VIM.new(vim_opts).tap do |vim|
-      vim.rev = vim.serviceContent.about.apiVersion
-      vim.serviceContent.sessionManager.Login(:userName => username, :password => password)
-    end
-
-    _log.info("Connected")
-    conn
-  end
-
-  def disconnect(vim)
-    return if vim.nil?
-
-    vim.close
-  end
-
-  def wait_for_updates(vim)
-    property_filter = create_property_filter(vim)
-
-    # Return if we don't receive any updates for 60 seconds break
-    # so that we can check if we are supposed to exit
-    options = RbVmomi::VIM.WaitOptions(:maxWaitSeconds => 60)
-
+  def initial_refresh
     # Send the "special initial data version" i.e. an empty string
     # so that we get all inventory back in the first update set
     version = ""
 
     # Use the full refresh persister for the initial UpdateSet from WaitForUpdates
     # After the initial UpdateSet this will change to a targeted persister
-    persister = ems.class::Inventory::Persister.new(ems)
-    parser    = ems.class::Inventory::Parser.new(persister)
+    persister = full_persister_klass.new(ems)
 
     _log.info("Refreshing initial inventory...")
 
-    initial = true
-    until exit_requested
-      update_set = vim.propertyCollector.WaitForUpdatesEx(:version => version, :options => options)
-      next if update_set.nil?
+    create_property_filter
 
-      # Save the new update set version
-      version = update_set.version
+    version = monitor_updates(version, persister)
 
-      process_update_set(property_filter, update_set, parser)
+    _log.info("Refreshing initial inventory...Complete")
 
-      next if update_set.truncated
+    version
+  end
 
-      save_inventory(persister)
+  def monitor_updates(version, persister = nil)
+    truncated = true
```

Comment:
I know why you did this... curious if just putting a `break` into the loop looks better here

Reply:
+1